src/porepy/numerics/ad/discretizations.py

```diff
@@ -479,166 +501,292 @@ def _differentiable_mpfa(
+                )
+                flux_jac += face_prolongation * vector_source_jac
 
         # Second part of product rule, applied to the potential. This is the standard
         # part of a Mpfa or Tpfa discretization
-        block_jac += base_flux * potential.jac
+        flux_jac += base_flux * potential.jac
 
         # The value of the flux is the standard mpfa/tpfa expression.
         block_val = base_flux * potential.val
 
-        return pp.ad.Ad_array(block_val, block_jac)
+        flux = pp.ad.Ad_array(block_val, flux_jac)
+        return flux
+
+    def _bound_pressure_face_function(
+        self,
+        perm_function: pp.ad.Function,
+        perm_argument: pp.ad.Ad_array,
+    ) -> pp.ad.Ad_array:
+        """The actual implementation of the
+
+        Parameters:
+            perm_function (pp.ad.Function): Function which gives the permeability as a
+                function of primary variables. The function should return cell-wise
+                permeability for all grids in grid_list.
+            perm_argument (pp.ad.Ad_array, evaluation of a pp.ad.Variable): Variable(s)
+                upon which the permeability depends.
+
+
+        Returns:
+            Ad_array: The flux, q, and its Jacobian matrix, where the latter accounts
+                for dependencies in the transmissibilities on cell center permeabilities.
+
+        """
+        # Note on parameters: When this function is called, potential should be an Ad
+        # operator (say, a MergedVariable representation of the pressure). During
+        # evaluation, because of the way operator trees are evaluated, potential will be
+        # an Ad_array (it is closer to being an atomic variable, thus it will be
+        # evaluated before this function).
+
+        # The product rule applied to q = T(k(u)) * p gives
+        #   dT/dk * dk/du * p + T * dp/dp.
+        # The first part is rather involved.
+
+        # Get hold of the underlying flux discretization.
+        base_bound_pressure_face = self._discretization.bound_pressure_face.evaluate(
+            self.dof_manager
+        )
+        # The Jacobian matrix should have the same size as the base.
+        bound_pressure_face_jac = sps.csr_matrix(
+            (base_bound_pressure_face.shape[0], self.dof_manager.num_dofs())
+        )
+
+        projections = pp.ad.SubdomainProjections(self.grids)
+
+        # The differentiation of transmissibilities with respect to permeability is
+        # implemented as a loop over all grids. It could be possible to gather all grid
+        # information in arrays as a preprocessing step, but that seems not to be worth
+        # the effort. However, to avoid evaluating the permeability function multiple
+        # times, we precompute it here
+        global_permeability = perm_function(perm_argument).evaluate(self.dof_manager)
+
+        g: pp.Grid
+        for g in self.grids:
+            params = self.gb.node_props(g)[pp.PARAMETERS][self.keyword]
```

Comment:
```suggestion
            params: Dict = self.gb.node_props(g)[pp.PARAMETERS][self.keyword]
```
I'm trying to make it a habit to always give type hints when accessing data from the data dictionaries. My idea is that this will be helpful for debugging, if anyone sends in something completely different at some point. I am a bit afraid that mypy at some point will start asking 'how do you know it is really a dict (or spmatrix etc.)?' - but for now, I believe this is a useful habit. Let me know what you think.

Reply:
Sounds sound!